File: src/gpx.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from xml.etree.ElementTree import Element, SubElement, indent, tostring

from src.models import Coordinate
from src.route import apply_route_jitter, build_timed_route


def _format_time(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def _prettify_xml(root: Element) -> str:
    indent(root, space="  ")
    return '<?xml version="1.0" encoding="utf-8"?>\n' + tostring(root, encoding="unicode")


def build_timed_points(
    points: list[Coordinate],
    speed_kmh: float = 5.0,
    jitter_meters: float = 0.0,
) -> list[tuple[Coordinate, datetime]]:
    timed_points = build_timed_route(points, speed_kmh=speed_kmh)
    return apply_route_jitter(timed_points, radius_meters=jitter_meters)


def _render_gpx(
    timed_points: list[tuple[Coordinate, datetime]],
    name: str = "GPS Simulation",
) -> str:
    root = Element(
        "gpx",
        {
            "version": "1.1",
            "creator": "Python iPhone GPS Simulator Prototype",
            "xmlns": "http://www.topografix.com/GPX/1/1",
        },
    )
    metadata = SubElement(root, "metadata")
    SubElement(metadata, "name").text = name
    SubElement(metadata, "time").text = _format_time(timed_points[0][1])

    if len(timed_points) == 1:
        point, timestamp = timed_points[0]
        waypoint = SubElement(root, "wpt", {"lat": f"{point.lat:.8f}", "lon": f"{point.lon:.8f}"})
        SubElement(waypoint, "ele").text = f"{point.elevation:.2f}"
        SubElement(waypoint, "time").text = _format_time(timestamp)
        SubElement(waypoint, "name").text = name
        return _prettify_xml(root)

    track = SubElement(root, "trk")
    SubElement(track, "name").text = name
    segment = SubElement(track, "trkseg")
    for point, timestamp in timed_points:
        track_point = SubElement(segment, "trkpt", {"lat": f"{point.lat:.8f}", "lon": f"{point.lon:.8f}"})
        SubElement(track_point, "ele").text = f"{point.elevation:.2f}"
        SubElement(track_point, "time").text = _format_time(timestamp)

    return _prettify_xml(root)


def build_gpx_points(points: list[Coordinate], speed_kmh: float = 5.0, jitter_meters: float = 0.0) -> list[Coordinate]:
    timed_points = build_timed_points(points, speed_kmh=speed_kmh, jitter_meters=jitter_meters)
    return [point for point, _timestamp in timed_points]


def generate_gpx(
    points: list[Coordinate],
    speed_kmh: float = 5.0,
    name: str = "GPS Simulation",
    jitter_meters: float = 0.0,
) -> str:
    timed_points = build_timed_points(points, speed_kmh=speed_kmh, jitter_meters=jitter_meters)
    return _render_gpx(timed_points, name=name)


def write_gpx_from_timed(
    path: str | Path,
    timed_points: list[tuple[Coordinate, datetime]],
    name: str = "GPS Simulation",
) -> Path:
    output_path = Path(path)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(_render_gpx(timed_points, name=name), encoding="utf-8")
    return output_path


def write_gpx(
    path: str | Path,
    points: list[Coordinate],
    speed_kmh: float = 5.0,
    jitter_meters: float = 0.0,
) -> Path:
    timed_points = build_timed_points(points, speed_kmh=speed_kmh, jitter_meters=jitter_meters)
    return write_gpx_from_timed(path, timed_points)

File: internal/core/gpx.go
package core

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

type gpxDocument struct {
	XMLName  xml.Name     `xml:"gpx"`
	Version  string       `xml:"version,attr"`
	Creator  string       `xml:"creator,attr"`
	XMLNS    string       `xml:"xmlns,attr"`
	Metadata gpxMetadata  `xml:"metadata"`
	Waypoint *gpxWaypoint `xml:"wpt,omitempty"`
	Track    *gpxTrack    `xml:"trk,omitempty"`
}

type gpxMetadata struct {
	Name string `xml:"name"`
	Time string `xml:"time"`
}

type gpxWaypoint struct {
	Lat       string `xml:"lat,attr"`
	Lon       string `xml:"lon,attr"`
	Elevation string `xml:"ele"`
	Time      string `xml:"time"`
	Name      string `xml:"name"`
}

type gpxTrack struct {
	Name    string     `xml:"name"`
	Segment gpxSegment `xml:"trkseg"`
}

type gpxSegment struct {
	Points []gpxTrackPoint `xml:"trkpt"`
}

type gpxTrackPoint struct {
	Lat       string `xml:"lat,attr"`
	Lon       string `xml:"lon,attr"`
	Elevation string `xml:"ele"`
	Time      string `xml:"time"`
}

func BuildTimedPoints(points []Coordinate, speedKmh, jitterMeters float64) ([]TimedPoint, error) {
	timed, err := BuildTimedRoute(points, speedKmh, time.Now().UTC(), time.Second)
	if err != nil {
		return nil, err
	}
	return ApplyRouteJitter(timed, jitterMeters, 0), nil
}

func GenerateGPX(points []Coordinate, speedKmh float64, name string, jitterMeters float64) (string, error) {
	timed, err := BuildTimedPoints(points, speedKmh, jitterMeters)
	if err != nil {
		return "", err
	}
	return RenderGPX(timed, name)
}

func RenderGPX(points []TimedPoint, name string) (string, error) {
	if len(points) == 0 {
		return "", fmt.Errorf("at least one timed point is required")
	}
	if name == "" {
		name = "GPS Simulation"
	}

	doc := gpxDocument{
		Version: "1.1",
		Creator: "Go iPhone GPS Simulator",
		XMLNS:   "http://www.topografix.com/GPX/1/1",
		Metadata: gpxMetadata{
			Name: name,
			Time: formatGPXTime(points[0].Time),
		},
	}

	if len(points) == 1 {
		point := points[0]
		doc.Waypoint = &gpxWaypoint{
			Lat:       fmt.Sprintf("%.8f", point.Point.Lat),
			Lon:       fmt.Sprintf("%.8f", point.Point.Lon),
			Elevation: fmt.Sprintf("%.2f", point.Point.Elevation),
			Time:      formatGPXTime(point.Time),
			Name:      name,
		}
	} else {
		trackPoints := make([]gpxTrackPoint, 0, len(points))
		for _, point := range points {
			trackPoints = append(trackPoints, gpxTrackPoint{
				Lat:       fmt.Sprintf("%.8f", point.Point.Lat),
				Lon:       fmt.Sprintf("%.8f", point.Point.Lon),
				Elevation: fmt.Sprintf("%.2f", point.Point.Elevation),
				Time:      formatGPXTime(point.Time),
			})
		}
		doc.Track = &gpxTrack{Name: name, Segment: gpxSegment{Points: trackPoints}}
	}

	var buffer bytes.Buffer
	buffer.WriteString(xml.Header)
	encoder := xml.NewEncoder(&buffer)
	encoder.Indent("", "  ")
	if err := encoder.Encode(doc); err != nil {
		return "", err
	}
	buffer.WriteByte('\n')
	return buffer.String(), nil
}

func WriteGPX(path string, points []TimedPoint, name string) error {
	rendered, err := RenderGPX(points, name)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(rendered), 0644)
}

func formatGPXTime(unixSeconds int64) string {
	return time.Unix(unixSeconds, 0).UTC().Format(time.RFC3339)
}
